File: app/src/main/java/com/temoa/gankio/ui/BaseFragment.java
package com.temoa.gankio.ui;

import android.os.Bundle;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

/**
 * Created by Temoa
 * on 2016/11/9 10:30
 */

public abstract class BaseFragment extends Fragment {

  private boolean hasCreateView = false;
  private boolean isFragmentVisible = false;

  protected View rootView;

  @Override
  public void setUserVisibleHint(boolean isVisibleToUser) {
    super.setUserVisibleHint(isVisibleToUser);

    if (rootView == null) {
      return;
    }

    hasCreateView = true;
    if (isVisibleToUser) {
      onFragmentVisibleChange(true);
      isFragmentVisible = true;
      return;
    }

    if (isFragmentVisible) {
      onFragmentVisibleChange(false);
      isFragmentVisible = false;
    }
  }

  @Override
  public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
    super.onViewCreated(view, savedInstanceState);
    if (!hasCreateView && getUserVisibleHint()) {
      onFragmentVisibleChange(true);
      isFragmentVisible = true;
    }
  }

  protected void onFragmentVisibleChange(boolean isVisible) {

  }
}

File: app/src/main/java/com/temoa/gankio/ui/BaseFragment.kt
package com.temoa.gankio.ui

import android.os.Bundle
import android.view.View
import androidx.fragment.app.Fragment

/**
 * Created by Temoa
 * on 2016/11/9 10:30
 */
abstract class BaseFragment : Fragment() {

  private var hasCreateView = false
  private var isFragmentVisible = false
  @JvmField
  protected var rootView: View? = null

  override fun setUserVisibleHint(isVisibleToUser: Boolean) {
    super.setUserVisibleHint(isVisibleToUser)
    if (rootView == null) {
      return
    }
    hasCreateView = true
    if (isVisibleToUser) {
      onFragmentVisibleChange(true)
      isFragmentVisible = true
      return
    }
    if (isFragmentVisible) {
      onFragmentVisibleChange(false)
      isFragmentVisible = false
    }
  }

  override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
    super.onViewCreated(view, savedInstanceState)
    if (!hasCreateView && userVisibleHint) {
      onFragmentVisibleChange(true)
      isFragmentVisible = true
    }
  }

  protected open fun onFragmentVisibleChange(isVisible: Boolean) {}
}